src/runtime/sysvar/slot_hashes.zig

```diff
@@ -1,42 +1,192 @@
+const builtin = @import("builtin");
 const std = @import("std");
 const sig = @import("../../sig.zig");
 
+const Allocator = std.mem.Allocator;
+
+const bincode = sig.bincode;
+
 const Hash = sig.core.Hash;
 const Pubkey = sig.core.Pubkey;
 const Slot = sig.core.Slot;
 
 /// [agave] https://github.com/anza-xyz/agave/blob/8db563d3bba4d03edf0eb2737fba87f394c32b64/sdk/slot-hashes/src/lib.rs#L43
 pub const SlotHashes = struct {
-    entries: []const Entry,
+    entries: std.ArrayListUnmanaged(Entry),
 
-    pub const Entry = struct { Slot, Hash };
+    pub const @"!bincode-config" = bincode.FieldConfig(SlotHashes){
+        .deserializer = deserialize,
+    };
 
-    pub const ID =
-        Pubkey.parseBase58String("SysvarS1otHashes111111111111111111111111111") catch unreachable;
+    pub const Entry = struct {
+        slot: Slot,
+        hash: Hash,
 
-    pub const DEFAULT = SlotHashes{
-        .entries = &.{},
+        pub fn sortCmp(_: void, a: Entry, b: Entry) bool {
+            return b.slot < a.slot; // Sort by descending slot
+        }
+
+        pub fn searchCmp(key: Slot, mid_item: Entry) std.math.Order {
+            return std.math.order(mid_item.slot, key);
+        }
     };
 
-    /// [agave] https://github.com/anza-xyz/solana-sdk/blob/834edeb5acf996377210729b0982819c42027227/sysvar/src/slot_hashes.rs#L59
-    pub const SIZE_OF: usize = 20_488;
+    pub const ID =
+        Pubkey.parseBase58String("SysvarS1otHashes111111111111111111111111111") catch unreachable;
 
-    // [agave] https://github.com/anza-xyz/solana-sdk/blob/9148b5cc95b43319f3451391ec66d0086deb5cfa/slot-hashes/src/lib.rs#L21
     pub const MAX_ENTRIES: usize = 512;
 
-    fn compareFn(key: Slot, mid_item: Entry) std.math.Order {
-        return std.math.order(key, mid_item[0]);
+    pub const SIZE_OF: usize = 20_488;
+
+    pub fn default(allocator: Allocator) Allocator.Error!SlotHashes {
```

Comment:
`default` is widely used in rust which is unfortunate because I don't think it's a great name for this kind of thing.

```suggestion
    pub fn empty(allocator: Allocator) Allocator.Error!SlotHashes {
```

A similar suggestion applies to the other sysvars

Reply:
For the allocated sysvars the default initialises them with the default max entries, and they are then not allowed to be resized. Perhaps using a BoundedArray containing an allocated slice would be preferable to enforce this constraint. As for the usage of default, I tend to agree with you and don't think its the best name either, however in this instance since it is associated with a fixed max size I think it kind of makes sense. Happy to run with empty if you prefer though, so lmk

Reply:
I think the bounded array idea makes more sense as well, would you be okay with me renaming these to `init` and transitioning these structs to use the bounded array in a follow up PR. They touch a lot of code and it may create more headaches with so many open dependent branches right now. Otherwise let me know and I will get straight on it now.

Reply:
I would prefer not to introduce technical debt, but if that overall makes the whole process easier, then the debt has negative interest, which sounds good to me, as long as it isn't forgotten about.

Reply:
where would you like to do this? in one of the stacked prs?

Reply:
what do you think about committing it in another pr and then we can resolve this?

Reply:
Yeah the plan would be to implement it on top of harnew/sysvars-with-stakes-cache, I will create a new PR now

Reply:
69ed3dc